keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -443,107 +453,173 @@ General prep:
+> Note: the delete workflow is described as synchronous, but it will likely run asynchronously to accommodate potentially long delete times when buckets contain many objects. The steps should still mostly follow what's outlined below.
 
-It is up to each provisioner whether or not to physically delete bucket content. But, for idempotency, the expectation is that the backend bucket will at least be made unavailable, pending out-of-band bucket lifecycle policies.
+Here is the delete workflow:
 
-The delete workflow is described as a synchronous, but it will likely be asynchronous to accommodate potentially long delete times when buckets contain many objects. The steps should still mostly follow what's outlined below.
++ User deletes their BR which is deferred until its finalizer has been removed.
++ COSI gets the `Bucket` instance from the BR and notes the deletion policy.
++ If the deletion policy is "_Retain_":
+   + COSI updates the `Bucket` status to "unavailable" or similar.
+   + COSI clears the BR reference in the `Bucket`.
+   + COSI removes all `Bucket` finalizers.
+      + Note: the `Bucket` instance and backend bucket are not deleted.
+      + Note: there may be pods and BAs referencing this `Bucket`.
+   + Note: at this point no new BARs are able to request access to this bucket.
 
-Here is the workflow:
++ If the deletion policy is "_Delete_":
+   + Note: the goal is to not delete the backend bucket or the `Bucket` instance until all access has been cleaned up.
+   + COSI deletes the `Bucket` instance which is deferred due to finalizers.
+      + Note: new BARs cannot access the `Bucket` now.
+   + The sidecar sees the deletionTimestamp set in the `Bucket` and checks to see if there are any BA.name finalizers in the `Bucket`.
+      + If none, it calls the driver to delete the backend bucket.
+         + COSI removes the "objectstorage.k8s.io/br-protect" finalizer from the `Bucket`, which allows it to be garbage collected.
+      + If >0, COSI does not call the driver and does no further processing.
+
++ If the deletion policy is "_ForceDeletea"_:
+   + COSI deletes the `Bucket` instance which is deferred due to finalizers.
+      + Note: new BARs cannot access the `Bucket` now.
+   + COSI gets all BAs referencing the `Bucket`.
+      + For each BA, COSI removes all finalizers and deletes the BA.
+      + Note: COSI does not delete any BARs that reference the deleted BAs, so a BAR may refer to a BA that no longer exists.
```

Comment:
Would it be wise to find the BARs that reference the deleted BA and set some status on them to indicate that the bucket they were using was force deleted?

Reply:
Yes. That actual status value will be defined by the implementation.